rustworkx-core/src/generators/random_graph.rs

```diff
@@ -395,10 +401,129 @@ where
     Ok(graph)
 }
 
+/// Generate a random Barabási–Albert preferential attachment algorithm
+///
+/// A graph of `n` nodes is grown by attaching new nodes each with `m`
+/// edges that are preferentially attached to existing nodes with high degree.
+/// If the graph is directed for the purposes of the extension algorithm all
+/// edges are treated as weak (meaning both incoming and outgoing).
+///
+/// The algorithm implemented in this function is described in:
+///
+/// A. L. Barabási and R. Albert "Emergence of scaling in random networks",
+/// Science 286, pp 509-512, 1999.
+///
+/// Arguments:
+///
+/// * `n` - The number of nodes to extend the graph to.
+/// * `m` - The number of edges to attach from a new node to existing nodes.
+/// * `seed` - An optional seed to use for the random number generator.
+/// * `initial_graph` - An optional starting graph to expand, if not specified
+///     a star graph of `m` nodes is generated and used. If specified the input
+///     graph is mutated by this function and is expected to be moved into this
+///     function.
+/// * `default_node_weight` - A callable that will return the weight to use
+///     for newly created nodes.
+/// * `default_edge_weight` - A callable that will return the weight object
+///     to use for newly created edges.
+///
+/// An `InvalidInput` error is returned under the following conditions. If `m < 1`
+/// or `m >= n` and if an `initial_graph` is specified and the number of nodes in
+/// `initial_graph` is `< m` or `> n`.
+///
+/// # Example
+/// ```rust
+/// use rustworkx_core::petgraph;
+/// use rustworkx_core::generators::barabasi_albert_graph;
+/// use rustworkx_core::generators::star_graph;
+///
+/// let graph: petgraph::graph::UnGraph<(), ()> = barabasi_albert_graph(
+///     20,
+///     12,
+///     Some(42),
+///     None,
+///     || {()},
+///     || {()},
+/// ).unwrap();
+/// assert_eq!(graph.node_count(), 20);
+/// assert_eq!(graph.edge_count(), 107);
+/// ```
+pub fn barabasi_albert_graph<G, T, F, H, M>(
+    n: usize,
+    m: usize,
+    seed: Option<u64>,
+    initial_graph: Option<G>,
+    mut default_node_weight: F,
+    mut default_edge_weight: H,
+) -> Result<G, InvalidInputError>
+where
+    G: Data<NodeWeight = T, EdgeWeight = M>
+        + NodeIndexable
+        + GraphProp
+        + NodeCount
+        + Build
+        + Create,
+    for<'b> &'b G: GraphBase<NodeId = G::NodeId> + IntoEdgesDirected + IntoNodeIdentifiers,
+    F: FnMut() -> T,
+    H: FnMut() -> M,
+    G::NodeId: Eq + Hash + std::fmt::Debug,
```

Comment:
Is there a reason for the Debug here?

Reply:
There isn't this is a leftover from when I was debugging the code. I'll remove it

Reply:
```suggestion
    G::NodeId: Eq + Hash,
```